srtcore/packetfilter.h

```diff
@@ -34,24 +34,88 @@ class PacketFilter
 
     typedef SrtPacketFilterBase* filter_create_t(const SrtFilterInitializer& init, std::vector<SrtPacket>&, const std::string& config);
 
+
+    // This is a filter container.
+    SrtPacketFilterBase* m_filter;
+    void Check()
+    {
+#if ENABLE_DEBUG
+        if (!m_filter)
+            abort();
+#endif
+        // Don't do any check for now.
+    }
+
+    // Filter is optional, so this check should be done always
+    // manually.
+    bool installed() const { return m_filter; }
+    operator bool() const { return installed(); }
+
+    SrtPacketFilterBase* operator->() { Check(); return m_filter; }
+
+    // In the beginning it's initialized as first, builtin default.
+    // Still, it will be created only when requested.
+    PacketFilter(): m_filter(), m_parent(), m_sndctlpkt(0), m_unitq() {}
+
+    // Copy constructor - important when listener-spawning
+    // Things being done:
+    // 1. The filter is individual, so don't copy it. Set NULL.
+    // 2. This will be configued anyway basing on possibly a new rule set.
+    PacketFilter(const PacketFilter& source SRT_ATR_UNUSED): m_filter(), m_parent(), m_sndctlpkt(0), m_unitq() {}
+
+    // This function will be called by the parent CUDT
+    // in appropriate time. It should select appropriate
+    // filter basing on the value in selector, then
+    // pin oneself in into CUDT for receiving event signals.
+    bool configure(CUDT* parent, CUnitQueue* uq, const std::string& confstr);
+
+    static bool correctConfig(const SrtFilterConfig& c);
+
+    // Will delete the pinned in filter object.
+    // This must be defined in *.cpp file due to virtual
+    // destruction.
+    ~PacketFilter();
+
+    // Simple wrappers
+    void feedSource(CPacket& w_packet);
+    SRT_ARQLevel arqLevel();
+    bool packControlPacket(int32_t seq, int kflg, CPacket& w_packet);
+    void receive(CUnit* unit, std::vector<CUnit*>& w_incoming, loss_seqs_t& w_loss_seqs);
+
+protected:
+    PacketFilter& operator=(const PacketFilter& p);
+    void InsertRebuilt(std::vector<CUnit*>& incoming, CUnitQueue* uq);
+
+    CUDT* m_parent;
+
+    // Sender part
+    SrtPacket m_sndctlpkt;
+
+    // Receiver part
+    CUnitQueue* m_unitq;
+    std::vector<SrtPacket> m_provided;
+};
+
+class PacketFilterFactory
```

Comment:
This rather should be a builtin class in PacketFilter and named something like "Global" or "Singleton" or something like that. Also all classes used there should be inside PacketFilter. This minimizes the introduced names and keeps the hierarchy flat.

This way you should be able to reach to this object by calling, e. g., `PacketFilter::glob().ParseConfig` etc.